fdtable: fix longstanding layering violations

cfriedt opened the pull request:
Although fdtable sat a layer below posix, as part of the base os, it was reaching up a layer for POSIX types, creating another dependency cycle.

Instead of forcing POSIX types to be used, simply add the native Zephyr types, `k_off_t` and `k_ssize_t`. Remove `<sys/types.h>` and `<fcntl.h>` inclusion below the POSIX layer, where appropriate.

Generally, it would be best to use only ISO or native zephyr types below POSIX.

(Would have fixed #75205 with a very small [workaround](https://gist.github.com/cfriedt/458a60d2827f230e709543fc33b73aa2))
Fixes #75513
Fixes #10436

Update:
To future-proof Zephyr in terms of file, disk, and device compatibility, a Kconfig option was added that allows the user to explicitly use a 64-bit `k_off_t`, even if the underlying hardware is only 32-bit.

> The question is whether we want future generations to pay off our debt + even more interest, or whether we are willing to pay the bill ourselves. This has already been ignored for 2 LTS releases.
> 
> Directly before an LTS release probably isn't a super appropriate time for a fix like this to go in, but then again, maybe it's the most appropriate time for it to go in 🤷‍♂️

cfriedt commented:
```shell
twister -p native_sim -p qemu_riscv64 -p qemu_x86 -p qemu_x86_64 -p qemu_cortex_a53 -p mps2/an385 -T tests/posix
...
INFO    - Total complete:  270/ 270  100%  skipped:   84, failed:    0, error:    0
INFO    - 45 test scenarios (270 test instances) selected, 84 configurations skipped (58 by static filter, 26 at runtime).
INFO    - 186 of 270 test configurations passed (100.00%), 0 failed, 0 errored, 84 skipped with 0 warnings in 403.48 seconds
INFO    - In total 7029 test cases were executed, 3759 skipped on 6 out of total 21 platforms (28.57%)
INFO    - 181 test configurations executed on platforms, 5 test configurations were only built.
```

aescolar commented on include/zephyr/sys/fdtable.h:
```diff
@@ -7,13 +7,33 @@
+ * e.g. <sys/types.h>, ssize_t, off_t.
+ */
+
+#ifndef _SSIZE_T_DECLARED
```
This guard is what pico (and newlib) use, but other libcs may have other guards (already minimal in tree)

cfriedt replied:
Can you point out a specific build that would be broken?

aescolar reviewed (dismissed):
Although I understand what and why you are doing this I don't think this is the solution.
Defining a type in the "wrong" header to avoid including another header from that one will end up with an even more messy situation, as now issues will depend on the order of header inclusion by users of those headers.
It is not safe to redefine these types (as you do not know what the C library is defining those to), and attempting to guard those definitions with the C library internal guard names is going to not work with other libraries or when those guards names change.

cfriedt commented:
@aescolar - this was the "path of least resistance" change. The better change would be to simply not use inappropriate types in the base OS (i.e. simply change return values from `ssize_t` to `long`). However, that would have greater repercussions as several function signatures in and out of tree would need to be corrected.

cfriedt commented:
Thanks for your help with this @keith-packard 🪁

Commit:
kernel: types: add k_off_t and k_ssize_t types

Define k_off_t to have native support for measuring file offsets
within any kind of file that is supportable in Zephyr.

CONFIG_OFFSET_64BIT is added to allow 32-bit users to force
64-bit offsets, when trading space and speed for flexibility
makes sense for the application.

Define k_ssize_t to have native support for reporting file I/O
sizes along with negative error codes.

Signed-off-by: Chris Friedt <cfriedt@tenstorrent.com>

Commit:
fdtable: remove inappropriate use of POSIX off_t and ssize_t

A shortcut was taken many years ago by "cherry-picking" the
POSIX off_t and ssize_t types for use throughout Zephyr.

Additionally, the POSIX header fcntl.h, as well as constants
defined in that header, were being used inappropriately
throughout Zephyr.

Doing so created a dependency cycle: below POSIX's position
in the stack, code depended on POSIX; above POSIX's position
in the stack, code depends on the lower parts.

Such dependency cycles usually result in fragility and
instability of the software stack.

Use the newly defined k_off_t and k_ssize_t types throughout
Zephyr's stack, where off_t and ssize_t were previously used
inappropriately.

Additionally, use ZVFS-prefixed constants instead of their
POSIX counterparts.

Additionally, where appropriate, ensure the POSIX fcntl.h
header is prefixed with <zephyr/posix/fcntl.h>.

We effectively create a mutual dependency to resolve the
cyclic dependency, as described in GitHub issue #51211.

For clarification, it is inappropriate to use POSIX types or
functions within the kernel, core OS, OS services, or really
anywhere that is not equal to or above POSIX's position in the
software stack.

In other words, if a library uses POSIX, Zephyr's POSIX
implementation cannot depend on that library.

Similarly, if a system service uses POSIX, Zephyr's POSIX
implementation cannot depend on that system service.

Signed-off-by: Chris Friedt <cfriedt@tenstorrent.com>

cfriedt commented:
Looks like there is 1 additional change to make.
Will check it out later when I am back at my desk.

```
seek and tell in file
[00:00:00.000,000] <err> fs: file seek error (-22)

    Assertion failed at ZEPHYR_BASE/tests/subsys/fs/common/test_fs_basic.c:164: seek_within_hello: (fs_seek(&file, -stat.size, FS_SEEK_CUR) not equal to 0)
verify hello seek back from cur failed
 FAIL - test_basic in 0.000 seconds
```

keith-packard commented on include/zephyr/types.h:
```diff
@@ -29,6 +29,36 @@ typedef union {
+ * k_ssize_t is tied to the word-size of the architecture. For compatibility purposes, k_ssize_t
+ * is defined as a 'long' on 64-bit architectures, and as an 'int' on 32-bit architectures.
+ */
+#ifdef CONFIG_64BIT
```
It seems like you explicitly need this to match the type of `size_t`, except signed? What newlib/picolibc do is a clever hack which works on both clang and gcc:
```
#ifdef __SIZE_TYPE__
/* If __SIZE_TYPE__ is defined (gcc) we define ssize_t based on size_t.
   We simply change "unsigned" to "signed" for this single definition
   to make sure ssize_t and size_t only differ by their signedness. */
#define unsigned signed
typedef __SIZE_TYPE__ _ssize_t;
#undef unsigned
#else
...
```
Stick your existing definitions inside the `#else` clause.

cfriedt replied:
Should be fixed now

keith-packard reviewed (changes requested):
As huge as this change is, I think it's probably the right direction. The only change I suggested is to use the hack newlib/picolibc have which creates a signed version of `size_t` from the gcc/clang definitions.

I will note that there are a bunch of changes here to code which also depend on other POSIX types, like `struct sockaddr`. Are you sure that code is "below" the POSIX layer?

cfriedt commented:
> I will note that there are a bunch of changes here to code which also depend on other POSIX types, like `struct sockaddr`. Are you sure that code is "below" the POSIX layer?

The other types were converted to native zephyr types in the gigantic revert of posix device io, posix fd mgmt, etc.

So had those been left in, they would be correct.

cfriedt commented:
Normally, my focus is fairly narrow in Zephyr, so I would never think to look in through (literally) the entire stack...

but I had no idea that the level of POSIX (ab)use was already at 11.

Nobody should be blamed, per se. In a lot of cases, `<sys/types.h>` was not even needed for inclusion (in many cases, it seemed to be used instead of `<stdint.h>` or `<stddef.h>`). In some cases, it was clear that an `int` was perfectly fine instead of `ssize_t` (e.g. for signed integers that would be maximally a few bytes).

I even added `ssize_t` by mistake for devmux, and there was a PR (that went stale) to fix it (#70633).

This is just ... really bad architecture. It's a dependency cycle at the API level, and is preventing critical features and stabilizing code from going in (see #75513). Code that will allow us to fix more bugs, more easily.

To find some semi-rational explanation, I assumed that there had must have been a TSC decision at some point on this topic.

I Googled "zephyr tsc meeting notes ssize_t off_t sys/types.h" and found this series of links

https://github.com/zephyrproject-rtos/zephyr/issues/10436
https://github.com/aescolar/zephyr/commit/d2258b02d15f0d6e2c7fb9473687fb3be16627f5

😞 🤦‍♂️ 

It was a bug reported in 2018... and it became stale and was closed in 2020, when the previous POSIX maintainer left the project (long before I became POSIX maintainer / Zephyr archaeologist).

There may have been a TSC decision on this topic, but Google & I cannot seem to find it.

In any case, regardless of whatever TSC decision was made, it's a dependency cycle (aka bug).

I've re-opened #10436.

Again, this is the correct solution as compared to the  "path of least resistance" [workaround](https://gist.github.com/cfriedt/458a60d2827f230e709543fc33b73aa2).

It seems we have found where the bodies were buried. It's a classic case of tech debt that has accrued a vast amount of interest. Undoubtedly, developers will defend the tech debt / decisions that were made, but that's usually just because their pride is damaged. The question is whether we want future generations to pay off our debt + even more interest, or whether we are willing to pay the bill ourselves. This has already been ignored for 2 LTS releases.

Directly before an LTS release probably isn't a super appropriate time for a fix like this to go in, but then again, maybe it's the most appropriate time for it to go in 🤷‍♂️

cfriedt commented:
* adjusted some print formatters
* added some casts
* rebased

cfriedt commented:
This is approximately a false positive, but should eventually be fixed in the fs subsystem.
```shell
subsys/fs/fcb/fcb.c:18:WARNING: Violation to rule 5.7 (Tag name should be unique) tag: fcb
```

cfriedt commented:
* Fixed / commented on weird doxygen issue in `retained_mem.h` due to strangely formatted build assert.

```
/*
 * Without the assert condition described this way, doxygen fails
 * with the error "retained_mem.h:187: unbalanced grouping commands".
 */
BUILD_ASSERT(!(sizeof(size_t) > sizeof(k_off_t)),
	     "Size of size_t must be equal or less than size of k_off_t");
```

cfriedt commented:
@sylvioalves - any idea why this change in type would be cause espressif builds to exhibit dram usage issues? Seems like it's specific to espressif. Have you seen something like this before?

https://github.com/zephyrproject-rtos/zephyr/actions/runs/9838628717/job/27159865954?pr=75348

```shell
/opt/toolchains/zephyr-sdk-0.16.8/xtensa-espressif_esp32_zephyr-elf/bin/../lib/gcc/xtensa-espressif_esp32_zephyr-elf/12.2.0/../../../../xtensa-espressif_esp32_zephyr-elf/bin/ld.bfd: zephyr/zephyr_pre0.elf section `.bss' will not fit in region `dram_seg'
/opt/toolchains/zephyr-sdk-0.16.8/xtensa-espressif_esp32_zephyr-elf/bin/../lib/gcc/xtensa-espressif_esp32_zephyr-elf/12.2.0/../../../../xtensa-espressif_esp32_zephyr-elf/bin/ld.bfd: DRAM segment data does not fit.
/opt/toolchains/zephyr-sdk-0.16.8/xtensa-espressif_esp32_zephyr-elf/bin/../lib/gcc/xtensa-espressif_esp32_zephyr-elf/12.2.0/../../../../xtensa-espressif_esp32_zephyr-elf/bin/ld.bfd: region `dram_seg' overflowed by 4104 bytes
collect2: error: ld returned 1 exit status
ninja: build stopped: subcommand failed.
```

sylvioalves commented:
> @sylvioalves - any idea why this change in type would be cause espressif builds to exhibit dram usage issues? Seems like it's specific to espressif. Have you seen something like this before?
> 
> https://github.com/zephyrproject-rtos/zephyr/actions/runs/9838628717/job/27159865954?pr=75348
> 
> ```shell
> /opt/toolchains/zephyr-sdk-0.16.8/xtensa-espressif_esp32_zephyr-elf/bin/../lib/gcc/xtensa-espressif_esp32_zephyr-elf/12.2.0/../../../../xtensa-espressif_esp32_zephyr-elf/bin/ld.bfd: zephyr/zephyr_pre0.elf section `.bss' will not fit in region `dram_seg'
> /opt/toolchains/zephyr-sdk-0.16.8/xtensa-espressif_esp32_zephyr-elf/bin/../lib/gcc/xtensa-espressif_esp32_zephyr-elf/12.2.0/../../../../xtensa-espressif_esp32_zephyr-elf/bin/ld.bfd: DRAM segment data does not fit.
> /opt/toolchains/zephyr-sdk-0.16.8/xtensa-espressif_esp32_zephyr-elf/bin/../lib/gcc/xtensa-espressif_esp32_zephyr-elf/12.2.0/../../../../xtensa-espressif_esp32_zephyr-elf/bin/ld.bfd: region `dram_seg' overflowed by 4104 bytes
> collect2: error: ld returned 1 exit status
> ninja: build stopped: subcommand failed.
> ```

I see.  Reason is that we added Espressif boards in `samples/sysbuild/with_mcuboot/sample.sysbuild.with_mcuboot` so we could guarantee build works. However, this PR somehow is causing that build to increase RAM usage. Let me check the root cause.

cfriedt commented:
@sylvioalves - it's possible that `k_off_t` being 64-bits (vs `off_t` previously being 32-bits on this arch) has added a few bytes to `.bss`.

sylvioalves commented:
@cfriedt I am checking why this is happening and will update it back soon. When I compare main build x this PR build, mcuboot increases 8k. That does not sum up at all.

sylvioalves commented on include/zephyr/storage/flash_map.h:
```diff
@@ -61,7 +62,7 @@ struct flash_area {
 	uint8_t fa_id;
 	uint16_t pad16;
 	/** Start offset from the beginning of the flash device */
-	off_t fa_off;
```
@cfriedt The reason for esp32* mcuboot to fail is due to this particular change. primary_slot_partition and secondary_slot_partition define at https://github.com/zephyrproject-rtos/mcuboot/blob/main/boot/bootutil/src/loader.c#L2095 are getting 2x bigger, from 4k to 8k. Is that the expected?

sylvioalves replied:
I get a similar increase when building for nrf52840dk/nrf52840 (@nordicjm)

v3.7.0-rc2:
```c
        Memory region         Used Size  Region Size  %age Used
           FLASH:       32830 B        48 KB     66.79%
             RAM:       23808 B       256 KB      9.08%
        IDT_LIST:          0 GB        32 KB      0.00%
```
This PR:
```
        Memory region         Used Size  Region Size  %age Used
           FLASH:       33406 B        48 KB     67.96%
             RAM:       25856 B       256 KB      9.86%
        IDT_LIST:          0 GB        32 KB      0.00%
        
 ```

sylvioalves replied:
In case this is expected (RAM increase), I'll update esp32 mcuboot builds accordingly.

sylvioalves commented:
> @sylvioalves - it's possible that `k_off_t` being 64-bits (vs `off_t` previously being 32-bits on this arch) has added a few bytes to `.bss`.

You mean there is no way to keep as is (32-bits) even with Kconfig stuff?

cfriedt commented:
> > @sylvioalves - it's possible that `k_off_t` being 64-bits (vs `off_t` previously being 32-bits on this arch) has added a few bytes to `.bss`.
> 
> You mean there is no way to keep as is (32-bits) even with Kconfig stuff?

We probably could make a Kconfig option for it. Previously, `off_t` was basically a `long`, so it would be 32-bit on 32-bit platforms and 64-bit on 64-bit platforms (similarly to `size_t`). 

I'm somewhat skeptical that there many people out there would explicitly require a 64-bit offset on a 32-bit platform, so a Kconfig would likely work.

E.g. `CONFIG_FLASH_MAP_OFFSET_64BIT=y`. Would need to be opt-in. Otherwise the default would be whatever the word size is for the architecture.

cfriedt commented:
* added `CONFIG_FLASH_MAP_OFFSET_64BIT` option (disabled by default)

@sylvioalves - even with the above, I'm still seeing 
```
section `.bss' will not fit in region `dram_seg'
region `dram_seg' overflowed by 4416 bytes
```
when testing with either of
```
west build --sysbuild -b esp32c3_devkitm samples/sysbuild/with_mcuboot
twister -i -p esp32c3_devkitm -s samples/sysbuild/with_mcuboot/sample.sysbuild.with_mcuboot
```

sylvioalves commented:
My conclusion is that anything defined using `k_off_t` is forcing it to be 64bit wide. It means structs like the below will duplicated the RAM usage. I guess some check is needed to keep it 32-bit wide..
```c
struct flash_area {
	/** ID number */
	uint8_t fa_id;
	uint16_t pad16;
	/** Start offset from the beginning of the flash device */
	k_off_t fa_off;
	/** Total size */
	size_t fa_size;
	/** Backing flash device */
	const struct device *fa_dev;
#if CONFIG_FLASH_MAP_LABELS
	/** Partition label if defined in DTS. Otherwise nullptr; */
	const char *fa_label;
#endif
};
```
```c
/**
 * @brief Structure for transfer flash sector boundaries
 *
 * This template is used for presentation of flash memory structure. It
 * consumes much less RAM than @ref flash_area
 */
struct flash_sector {
	/** Sector offset from the beginning of the flash device */
	k_off_t fs_off;
	/** Sector size in bytes */
	size_t fs_size;
};
```

cfriedt commented:
Fixed with your solution (offline). Thanks for your help @sylvioalves 🪁 🙏

sylvioalves commented on include/zephyr/types.h:
```diff
@@ -29,6 +29,40 @@ typedef union {
+ * In order to overcome historical limitations of the off_t type, Zephyr always defines
+ * k_off_t as a 64-bit signed integer.
+ */
+typedef int64_t k_off_t;
```
Every `off_t` is changed to `k_off_t`, meaning the system footprint will increase. Is this really correct?

cfriedt replied:
Just for illustration, I would normally expect to see `k_off_t` used in thread stacks.

If an application used a new `k_off_t` in every function (which would be unusual), and the stack was 25 function calls deep (which is an exceptionally deep stack), then there would only be 100 bytes of additional stack space used. That average case would be significantly less.

So not too bad.

The benefit is, of course, that Zephyr would always be guaranteed to handle any practically sized file, even if it's 20 GiB or something crazy like that.

cfriedt replied:
Moved the Kconfig option to the kernel and renamed `CONFIG_OFFSET_64BIT`, much like `CONFIG_TIMEOUT_64BIT`.

It is not enabled by default, but defers the decision to the user.

At least that sets us up for future-proofing `k_off_t`, @keith-packard.

keith-packard replied:
yeah, I get excited when I can reduce a function by four bytes in size... Making the shift from a POSIX type to a Zephyr type is sufficient to resolve the actual bug, we can argue when that type should shift to 64 bits in some later PR. Good call.

nordicjm replied:
am I missing where the size is set? This is 64-bit always, which really is not needed for the vast majority of low flash devices zephyr supports

cfriedt replied:
Fixed!

cfriedt commented:
* Fix `Violation to rule 5.7 (Tag name should be unique) tag: fcb` by renaming function arguments
* Provided a Kconfig option to force `k_off_t` to 64-bits (by default, it matches the architecture word size)

ycsin reviewed (commented):
Very much in favor of this PR going into the 3.7 LTS so that applications/drivers/etc developing on that release would be using the correct type from the start.

cfriedt commented on kernel/Kconfig:
```diff
@@ -866,6 +866,23 @@ config XIP
 	  supply a linker command file when building your image. Enabling this
 	  option increases both the code and data footprint of the image.
 
+config OFFSET_64BIT
```
Thanks for finding this

cfriedt replied:
Fixed!

nashif commented:
@cfriedt this change falls under [treewide changes](https://docs.zephyrproject.org/latest/contribute/guidelines.html#treewide-changes) and needs to follow the process for such changes. (see link).

cc: @carlescufi

nashif commented:
> Fixes #75513
> Fixes #10436

isnt #75513 duplicating #10436 now? Can we please close one of them?

cfriedt commented:
> isnt #75513 duplicating #10436 now? Can we please close one of them?

They should both be closed when this PR is merged.

nashif commented:
> They should both be closed when this PR is merged.

that I am aware of, but if both issues report the same problem, we should track just one with one priority, if they are two different issues, we can track them both.

cfriedt commented:
> > They should both be closed when this PR is merged.
> 
> that I am aware of, but if both issues report the same problem, we should track just one with one priority, if they are two different issues, we can track them both.

I don't know if they both report exactly the same problem. They are definitely unique. The more recent bug report includes a lot more detail that is relevant *now*. 

It seems like closing either one of them would be like sweeping dirt under the carpet. Why bother? There is nothing to hide.

cfriedt commented:
> @cfriedt this change falls under [treewide changes](https://docs.zephyrproject.org/latest/contribute/guidelines.html#treewide-changes) and needs to follow the process for such changes. (see link).
> 
> cc: @carlescufi

Done #75513

cc: @carlescufi

carlescufi commented:
> > @cfriedt this change falls under [treewide changes](https://docs.zephyrproject.org/latest/contribute/guidelines.html#treewide-changes) and needs to follow the process for such changes. (see link).
> > cc: @carlescufi
> 
> Done #75513
> 
> cc: @carlescufi

I had to cancel the meeting today @cfriedt, sorry about that.

cfriedt commented:
> > > @cfriedt this change falls under [treewide changes](https://docs.zephyrproject.org/latest/contribute/guidelines.html#treewide-changes) and needs to follow the process for such changes. (see link).
> > > cc: @carlescufi
> > 
> > 
> > Done #75513
> > cc: @carlescufi
> 
> I had to cancel the meeting today @cfriedt, sorry about that.

No worries @carlescufi.

Commit:
manifest: optional: update sof to use k_ssize_t

Update sof to use k_ssize_t instead of the POSIX ssize_t, thus
avoiding layering violations in core parts of Zephyr.

Signed-off-by: Chris Friedt <cfriedt@tenstorrent.com>

de-nordic commented on drivers/hwinfo/hwinfo_gecko.c:
```diff
@@ -19,7 +19,7 @@
 #endif
 
-ssize_t z_impl_hwinfo_get_device_id(uint8_t *buffer, size_t length)
+int z_impl_hwinfo_get_device_id(uint8_t *buffer, size_t length)
```
same here

cfriedt replied:
No change then?

de-nordic replied:
We go with int and cry.

nashif commented:
> I had to cancel the meeting today @cfriedt, sorry about that.

we can talk about this in the release WG meeting later today.

nashif commented:
> It seems like closing either one of them would be like sweeping dirt under the carpet. Why bother? There is nothing to hide.

closing duplicate issues and focusing on one issue is not sweeping dirt under the carpet. We have always done this and this is not new. Closing a duplicate issue does not make it less important. Nobody is trying to HIDE anything.

de-nordic commented on drivers/disk/flashdisk.c:
```diff
@@ -28,13 +28,13 @@ struct flashdisk_data {
 	struct k_mutex lock;
 	const unsigned int area_id;
-	const off_t offset;
+	const k_off_t offset;
```
I am not sure whether using `off_t` or `k_off_t` is correct here.
I guess that the `off_t` has been used here because it may have larger range than `size_t`, but using signed int here is pointless, same with `cached_addr`. Note also that using 64bit values here, just to serve device that is >2GB and <=4GB, means that upper 4 bytes of the `off_t` will never be used.

I think that we need separate types for storage/external devices, with long addresses, independent from whatever kernel is running on.

cfriedt replied:
`k_off_t` (and `off_t`) are specifically for describing sizes and offsets within storage devices.

Either is fine with me, but this PR is specifically addressing the layering violation.

It might be better to retype this field in a separate PR.

cfriedt commented on drivers/hwinfo/hwinfo_esp32.c:
```diff
@@ -13,7 +13,7 @@
 #include <string.h>
 
-ssize_t z_impl_hwinfo_get_device_id(uint8_t *buffer, size_t length)
+int z_impl_hwinfo_get_device_id(uint8_t *buffer, size_t length)
```
This function could fail if a parameter were invalid.

But then again, it should probably also fail if the device ID exceeds `INT32_MAX` 😅.

It's probably safe.

de-nordic replied:
I hate `size_t`, it is pointless type. `sizeof` should be returning `unsigned int`.

de-nordic replied:
The C standard only states that `size_t` should have conversion rank not greater than `signed long` (+ the implementation specific magic words), so it does not seem that there is easy way to find out what is best arithmetic type equivalent. So `int` it is.

cfriedt replied:
`size_t` is standard C (not POSIX), and is used pretty ubiquitously this way. It's not really a concern in this PR which is just addressing the layering violation. If `int` is otherwise fine, should this comment be marked as resolved?

de-nordic replied:
> `size_t` is standard C (not POSIX), and is used pretty ubiquitously this way. It's not really a concern in this PR which is just addressing the layering violation. If `int` is otherwise fine, should this comment be marked as resolved?

Yes, I know. It just shows how the language standard is out of touch with reality.

cfriedt commented:
* updated `west.yml` to point at sof PR

cfriedt commented:
This appears to be passing CI now and should be merged before v3.7.0

nashif commented:
> This appears to be passing CI now and should be merged before v3.7.0

@cfriedt  the nature of this change which involves changing types, touching 400 files and changing sginatures of many APIs requires more discussion and following the process outlined above, i.e. discussion and consesnus in architecture working group and wide broadcasting of the change to the community and those affected. 

This would have been the case if we were not in the middle of a release. The fact that we are in the middle of a release and closing on a hard freeze milestone in 2 days makes it a bit more critical to get a signoff from stakeholders, not only rely on CI passing to get it merged.

The arch. wg was cancelled yesterday, so I can give you a slot in today's TSC to make the case for this to go into 3.7, if you wish to do that please mark this issue with the TSC label and join the TSC today.

cfriedt commented:
@nashif - I have an appointment during the TSC call unfortunately. It's not something I usually plan on attending these days.

Additionally, I would be advocating for the same things I advocated for the last time I attended the TSC meeting. I.e. putting v3.7-branch into a better state for additional bug fixes and stabilization fixes.

At this point, it's unlikely that POSIX will be able to receive any significant backports in v3.7-branch.

There are also other fairly significant bugs that would be part of the release at this point (such as #75402).

Personally, I wouldn't consider tagging v3.7.0 without additional bug fixes, but that isn't my decision to make.

ycsin reviewed (dismissed):
+1 to fixing the layering issue, it is unfortunate that there are a lot of affected files and we didn't get to stop this from snowballing much earlier.

It is pretty late in the current release cycle, but for users that jump from one LTS release to the next, getting this sorted out before the release means that LTS v3 users will be better positioned for the next LTS release.

kartben commented:
> It is pretty late in the current release cycle, but for users that jump from one LTS release to the next, getting this sorted out before the release means that LTS v3 users will be better positioned for the next LTS release.

Not only is it late in the cycle but it still is a tree-wide change so I believe the process linked in a previous comment would still need to be followed, and this potential change communicated to a wider audience ASAP.

cfriedt commented:
> Not only is it late in the cycle but it still is a tree-wide change so I believe the process linked in a previous comment would still need to be followed, and this potential change communicated to a wider audience ASAP.

Yes, the process requires the arch wg to approve, and then when it is stamped, an email should be sent out to the devel mailing list before it is merged, etc. If the arch wg does not come to a consensus it goes to the TSC, etc.

For the arch wg to approve, there would need to be consensus, and I think that would be the challenging part; some people really like defending bad decisions.

It would also need some buy-in from the release manager... but... 🤷🏼‍♂️

It is very late in the release cycle for all of this.

Of course; this should have been fixed many, many, releases ago 🤷🏼‍♂️

cfriedt commented:
* rebased on `main` after #75654 was merged

sylvioalves reviewed (approved):
From ESP32 side and memory discussion, LGTM.

andyross reviewed (commented):
Ooph, that's a lot of change.  Might as well +1 just so we can get this in.  In the longer term, maybe we should look at moving usage from ssize_t (AIUI a POSIX type) to ptrdiff_t (ISO C).  Not sure whether off_t has an equivalent in the standard library, though?   Probably not, so I guess a k_off_t is unavoidable.

andyross reviewed (approved):
Oops forgot vote

jhedberg commented on drivers/bluetooth/hci/userchan.c:
```diff
@@ -188,7 +188,7 @@ static void rx_thread(void *p1, void *p2, void *p3)
 		size_t buf_tailroom;
 		size_t buf_add_len;
-		ssize_t len;
+		k_ssize_t len;
```
Is this really correct? This driver is for native_sim, and some 10 lines later the return value from the host libc's `read()` is assigned to it.

cfriedt replied:
The definitions should be identical.

jhedberg replied:
I don't doubt that. I was mainly wondering where a `native_sim` driver like this conceptually sits in the software stack. It's calling the host OS APIs (like `read()`) so in that sense it could be seen to be above POSIX (in which case `ssize_t` would be more appropriate), however the driver also accesses Zephyr kernel APIs. Anyway, this isn't necessarily that critical (as you say, the types should be the same), which is why I didn't want to block this PR by requesting changes, but I do think it would be good to understand where `native_sim` driver code sits in the software stack.

cfriedt replied:
Yea, it's kind of a recursion. I would still say those drivers sit below (Zephyr's) POSIX API.

The namespace issue is kind of an unfortunate side-effect (not without solutions, ofc)

andyross commented:
> Not sure whether off_t has an equivalent in the standard library, though?

Duh, I forgot stdio.  So interestingly fread/fwrite() use `size_t` as the buffer offset type, so effectively in all real circumstances that demands off_t be the same.  Interestingly fseek() never got the memo and uses a bare `long` as the offset type.  But there, POSIX got huffy and defined `fseeko()` which takes off_t, so we don't get to assume anything.

cfriedt commented:
Technically size_t would be the same already for 64 bit.

keith-packard commented:
> Duh, I forgot stdio. So interestingly fread/fwrite() use `size_t` as the buffer offset type, so effectively in all real circumstances that demands off_t be the same.

Not at all -- `size_t` refers to the in-memory storage, which is limited to the address space. `off_t` covers file sizes, not memory, which C doesn't really talk about much, aside from `fseek` as you say. However, there's also `fpos_t` as used by `fgetpos` and `fsetpos`. C has weasel words for that allowing it to be a struct type, which allows file sizes beyond the scope of any integer type, but sadly doesn't map neatly to `off_t`.

On 32-bit Linux systems, `size_t` is 32-bits, but (most of the time), `off_t` is 64-bits.

cfriedt commented:
Aren't standards wonderful?? 😂🙃

nashif commented:
The PR is reaching the limit of what GH can handle, opening it in the browser just takes forever :(

from the commit message of the largest commit:

> Additionally, the POSIX header fcntl.h, as well as constants
    defined in that header, were being used inappropriately
    throughout Zephyr.

Can this be done in its own commit?

>     Additionally, use ZVFS-prefixed constants instead of their
    POSIX counterparts.

ditto

>     Additionally, where appropriate, ensure the POSIX fcntl.h
    header is prefixed with <zephyr/posix/fcntl.h>.

ditto

nashif commented on kernel/Kconfig:
```diff
@@ -866,6 +866,23 @@ config XIP
 	  option increases both the code and data footprint of the image.
 
+config OFFSET_64BIT
+	bool "Use 64-bit offsets"
```
would be great if this went into its own section in Kconfig (menu), i.e. do not put alongside other features and make it obvious is it a type, maybe others will be added to the same menu later.

`OFFSET_64BIT` can be confusing, maybe make it obvious we are talking about a type here: `TYPE_OFFSET_64BIT`.

cfriedt replied:
Sure. The closest thing I found to this so far is `TIMEOUT_64BIT`. Should that be moved as well? It also has to do with a type (`k_timeout_t`).

de-nordic commented:
> > Not sure whether off_t has an equivalent in the standard library, though?
> 
> Duh, I forgot stdio. So interestingly fread/fwrite() use `size_t` as the buffer offset type, so effectively in all real circumstances that demands off_t be the same. Interestingly fseek() never got the memo and uses a bare `long` as the offset type. But there, POSIX got huffy and defined `fseeko()` which takes off_t, so we don't get to assume anything.

`fread/fwrite` operate on memory objects rather than buffers. The POSIX `write` operates on buffers, but still knows it can only write as much as `size_t`.
I think that when the `fread/fwrite` was designed you would still mostly write on tape, so you would rather do a lot of sequential operations, on same types of objects, and then `fseek` to some other location, and again do some sequential operations.
Tape could be much longer then your address space so using long for `fseek` made sense, as this operates on external address space which may not fit in `size_t`.
POSIX was designed later than C, to stop Sys V vs 4BSD API division in systems getting to annoying.
It is funny that if you have true POSIX system, this means that `lseek` would be native function using `off_t`, for offset, and the `fseek` implemented on top would have to convert from `long` to `off_t` on call.

mmahadevan108 commented:
@cfriedt, can you help rebase.
